pkg/agent/storage/azure.go

```diff
@@ -0,0 +1,217 @@
+//   - prefix (the path to the blob within the containerName) would be 'this/is/virtualDir/fileName'.
+func parseAzureUri(uri string) (azureUriParts, error) {
+	if !strings.HasPrefix(uri, string(AZURE)) {
+		return azureUriParts{}, fmt.Errorf("invalid Azure URI: '%s'. The URI must start with '%s'", uri, string(AZURE))
+	}
+
+	uri = strings.TrimPrefix(uri, string(AZURE))
+	tokens := strings.SplitN(uri, "/", 3)
+	if len(tokens) < 3 {
+		return azureUriParts{}, fmt.Errorf("invalid Azure URI: '%s'. The URI must be in the format 'azure://<serviceUrl>/<containerName>/<blob>'", uri)
+	}
+
+	serviceUrl := tokens[0]
+	containerName := tokens[1]
+	prefix := ""
+	if len(tokens) == 3 {
+		prefix = tokens[2]
+	}
+
+	return azureUriParts{
+		serviceUrl:    string(HTTPS) + serviceUrl,
+		containerName: containerName,
+		prefix:        prefix,
+	}, nil
+}
+
+// azureUriParts should follow this specific format: azure://<serviceUrl>/<containerName>/<blob>
+type azureUriParts struct {
+	serviceUrl    string
+	containerName string
+	prefix        string
+}
+
+type defaultAzureDownloader struct {
+	modelDir      string
+	modelName     string
+	containerName string
+	prefix        string
+	client        *azblob.Client
+}
+
+var _ AzureDownloader = (*defaultAzureDownloader)(nil)
+
+func (d *defaultAzureDownloader) SetDownloaderValues(modelDir string, modelName string, container string, prefix string) {
+	d.modelDir = modelDir
+	d.modelName = modelName
+	d.containerName = container
+	d.prefix = prefix
+}
+
+func (d *defaultAzureDownloader) GetAllObjects() ([]string, error) {
+	var results []string
+	pager := d.client.NewListBlobsFlatPager(d.containerName, &container.ListBlobsFlatOptions{Prefix: &d.prefix})
+	for pager.More() {
+		page, err := pager.NextPage(context.TODO())
+		if err != nil {
+			return nil, err
+		}
+		for _, blob := range page.Segment.BlobItems {
+			results = append(results, *blob.Name)
+		}
+	}
+
+	return results, nil
+}
+
+func (d *defaultAzureDownloader) Download(blobsNames []string) error {
+	for _, blobName := range blobsNames {
+		err := d.DownloadSingle(blobName)
+		if err != nil {
+			return fmt.Errorf("failed to download blob '%v': %v", blobName, err)
+		}
+	}
+
+	return nil
+}
+
+func (d *defaultAzureDownloader) DownloadSingle(blobName string) error {
+	relativeFilePath := strings.TrimPrefix(blobName, d.prefix)
+	filePath := filepath.Join(d.modelDir, d.modelName, relativeFilePath)
+
+	if FileExists(filePath) {
+		// File got corrupted or is mid-download :(
+		if err := os.Remove(filePath); err != nil {
+			return fmt.Errorf("file is unable to be deleted: %w", err)
+		}
+	}
+
+	file, err := Create(filePath)
+	defer func(file *os.File) {
+		err := file.Close()
+		if err != nil {
+			log.Error(err, "file is unable to be closed")
+		}
+	}(file)
+	if err != nil {
+		return fmt.Errorf("file is already created: %w", err)
+	}
+	_, err = d.client.DownloadFile(context.TODO(), d.containerName, blobName, file, &azblob.DownloadFileOptions{})
```

Comment:
what happens if the target azure blob is protected by auth? (same for getAllObjects)

Reply:
Azure client reads the environment variables which contain tenant id, client id and client secret. The same set of variables as in https://github.com/leduckhc/kserve/blob/master/pkg/credentials/azure/azure_secret.go#L33